src/ATen/templates/Type.h

```diff
@@ -37,6 +37,8 @@ struct Generator;
 // situation.
 constexpr int64_t kUndefinedDimensions = std::numeric_limits<int64_t>::min();
 
+static void noop_deleter(void*) {}
```

Comment:
the destructor signature used in dlpack is like `void (*)(void *, void*)` because of callback
do we need to change the noop_deleter signature as well to be consistent?

Reply:
The DLPack deleter is pure C, so it needs a void * ctx pointer. The ATen API is C++ so it takes a std::function<void(void*>>. This object can be a C++ closure (created via C++'s lambda functions), which can store the ctx pointer.
```
Tensor fromDLPack(const DLManagedTensor* src) {
  Backend backend = getATenBackend(src->dlTensor.ctx);
  ScalarType stype = getATenScalarType(src->dlTensor.dtype);
  auto deleter = [src](void * data) {
      src->destructor(src, src->ctx);
  };
  return getType(backend, stype).tensorFromBlob(
      src->dlTensor.data,
      IntList(src->dlTensor.shape, src->dlTensor.ndim),
      IntList(src->dlTensor.strides, src->dlTensor.ndim),
      deleter);
}
```

Reply:
We can also change the DLManagedTensor object to:

```
typedef struct {
  /*! \DLTensor which is being memory managed */
  DLTensor dlTensor;
  /*! \brief context in which DLManagedTensor is used in a framework */
  void * ctx;
   /*! \brief Destructor signature void (*)(void*, void*) - this should be called
    *   to destruct ctx which holds the DLManagedTensor. It can be NULL if there
    *   is no way for the caller to provide a reasonable destructor.
    */
   void (*destructor)(DLManagedTensor * self);
 } DLManagedTensor;
```
Because you can access the ctx object from `self`, so there is no need for it to be the second argument. 

I am pretty sure you were suggesting this the otherday, but I didn't understand until now.

Reply:
awesome! that makes much more sense now, it was driving me crazy earlier